std/container/binaryheap.d

```diff
@@ -369,12 +369,40 @@ must be collected.
             insert(value);
             return true;
         }
-        // must replace the top
+
+        assert(!_store.empty, "Cannot replace front of an empty heap.");
+        if (!comp(value, _store.front)) return false; // value >= largest
+        _store.front = value;
+
+        percolate(_store[], 0, _length);
+        debug(BinaryHeap) assertValid();
+        return true;
+    }
+
+/**
+If the heap has room to grow, inserts $(D value) into the store and
+returns $(D true). Otherwise, if $(D less(value, front)), copies front
+in value and returns $(D true). Otherwise, leaves
+the heap unaffected and returns $(D false).
+     */
+    bool conditionalSwap(ref ElementType!Store value)
+    {
+        _payload.refCountedStore.ensureInitialized();
+        if (_length < _store.length)
+        {
+            insert(value);
+            return true;
+        }
+
         assert(!_store.empty, "Cannot replace front of an empty heap.");
+        auto frontCopy = _store.front;
         if (!comp(value, _store.front)) return false; // value >= largest
         _store.front = value;
+        value = frontCopy;
```

Comment:
The point here is to use the swap primitive instead of assignment (it's cheap and nothrow). So the code would go:

```d
 if (!comp(value, _store.front)) return false; // value >= largest
swap(_store.front, value);
```